### PR TITLE
Update GPT2 I/O definition to be recognized by ORT optimizer.

### PR DESCRIPTION
**Pull request by mfuntowicz**
This PR aims at making the GPT2 + past more compatible with ONNX Runtime optimizer (especially attention fusion).
Past key values should be presented as a single tensor with both key and value stacked on the leading axis.

It also provide a concat mecanism to merge the resulting past key values so it's a single tensor too.

- [x] I/O shapes changes
- [x] I/O dtype changes
- [ ] Past keys wrapper
- [ ] Monkey Patching
- [ ] Unittests

### CODE REVIEW

File: src/transformers/models/gpt2/configuration_gpt2.py [hunk 1]
```diff
@@ -249,25 +249,25 @@ def generate_dummy_inputs(
 
                 batch, seqlen = common_inputs["input_ids"].shape
                 # Not using the same length for past_key_values
-                past_key_values_length = seqlen + 2
+                past_key_values_length = 1
```
Comment (michaelbenayoun) [inline src/transformers/models/gpt2/configuration_gpt2.py:252]: Why?

File: src/transformers/models/gpt2/configuration_gpt2.py [hunk 2]
```diff
@@ -249,25 +249,25 @@ def generate_dummy_inputs(
...
                 past_shape = (
+                    2,
                     batch,
                     self.num_attention_heads,
                     past_key_values_length,
                     self._config.hidden_size // self.num_attention_heads,
                 )
                 ordered_inputs["past_key_values"] = [
-                    (torch.zeros(past_shape), torch.zeros(past_shape)) for _ in range(self.num_layers)
+                    torch.zeros(past_shape, dtype=torch.float32) for _ in range(self.num_layers)
```
Comment (michaelbenayoun) [inline src/transformers/models/gpt2/configuration_gpt2.py:261]: Does it work with latest `transformers` versions?
Because the expecte input is a tuple of tuples of tensors. It might work because indexing would behave the same way?

File: src/transformers/models/gpt2/configuration_gpt2.py [hunk 3]
```diff
@@ -249,25 +249,25 @@ def generate_dummy_inputs(
...
 
-        ordered_inputs["attention_mask"] = common_inputs["attention_mask"]
-        if self.use_past:
-            ordered_inputs["attention_mask"] = torch.cat(
-                [ordered_inputs["attention_mask"], torch.ones(batch, past_key_values_length)], dim=1
-            )
+            if self.use_past:
+                ordered_inputs["attention_mask"] = torch.ones(batch, seqlen + past_key_values_length, dtype=torch.int64)
+        else:
+            ordered_inputs["attention_mask"] = common_inputs["attention_mask"].long()
```
Comment (michaelbenayoun) [inline src/transformers/models/gpt2/configuration_gpt2.py:267]: Do we need to cast?

File: src/transformers/onnx/config.py [hunk 4]
```diff
@@ -418,8 +418,9 @@ def generate_dummy_inputs(
 
             batch, seqlen = common_inputs["input_ids"].shape
             # Not using the same length for past_key_values
-            past_key_values_length = seqlen + 2
```
Comment (michaelbenayoun) [inline src/transformers/onnx/config.py:421]: Why?

File: src/transformers/onnx/config.py [hunk 5]
```diff
@@ -452,12 +453,10 @@ def fill_with_past_key_values_(self, inputs_or_outputs: Mapping[str, Mapping[int
 
         name = "past_key_values" if direction == "inputs" else "present"
         for i in range(self.num_layers):
-            inputs_or_outputs[f"{name}.{i}.key"] = {0: "batch", 2: "past_sequence + sequence"}
-            inputs_or_outputs[f"{name}.{i}.value"] = {0: "batch", 2: "past_sequence + sequence"}
+            inputs_or_outputs[f"{name}.{i}"] = {1: "batch", 3: "past_sequence + sequence"}
```
Comment (michaelbenayoun) [inline src/transformers/onnx/config.py:456]: Nice catch!

**Review by michaelbenayoun — COMMENTED** [review]
Great!
Left a few comments and questions (some relevant, some less).

Have you tried running this?
```
RUN_SLOW=1 pytest tests/onnx/test_onnx_v2.py
```

**Review by michaelbenayoun — APPROVED** [review]

File: src/transformers/models/gpt2/configuration_gpt2.py (hunk 2 above)
Comment (lewtun) [reply src/transformers/models/gpt2/configuration_gpt2.py:261]: Indeed, it would be nice to see an explicit example with a GPT-2 checkpoint, with `past_key_values` passed to the model

File: src/transformers/onnx/config.py [hunk 6]
```diff
@@ -246,6 +243,17 @@ def _generate_dummy_images(
             images.append(Image.fromarray(data.astype("uint8")).convert("RGB"))
         return images
 
+    def install_patching_spec(self, spec: PatchingSpec):
+        """
+        Append a patching spec after initialization of the configuration
+        :param spec:
```
Comment (lewtun) [inline src/transformers/onnx/config.py:249]: Could we use the same docstring style that's used in this class?

E.g.

```
"""
Append a patching spec after initialization of the configuration

Args:
    spec: The patching spec.
"""
```

**Review by lewtun — APPROVED** [review]
Thanks a lot for giving GPT-2 a much needed fix for past key values! Apart from a nit about the docstrings and @michaelbenayoun's comments, this LGTM 🔥
